FIAT/macro.py

```diff
@@ -282,43 +247,74 @@ def construct_subcomplex(self, dimension):
 
-                # Connect subentity barycenter to the entity barycenter
-                for subdim in range(dim):
-                    edges.extend((inv_top[subdim][subverts] + offset[subdim], bary_id)
-                                 for subverts in combinations(top[dim][entity], subdim+1))
-
-        new_topology = make_topology(sd, len(new_verts), edges)
-        parent = ref_el.get_parent() or ref_el
-        super(PowellSabinSplit, self).__init__(parent, tuple(new_verts), new_topology)
+                # Connect entity barycenter to every subsimplex on the entity
+                simplices[dim][entity] = [(*s, bary_id)
+                                          for child in connectivity[(dim, dim-1)][entity]
+                                          for s in simplices[dim-1][child]]
+
+        simplices = list(chain.from_iterable(simplices[sd].values()))
+        new_topology = {}
+        new_topology[0] = {i: (i,) for i in range(len(new_verts))}
+        for dim in range(1, sd):
+            facets = chain.from_iterable((combinations(s, dim+1) for s in simplices))
+            if dim < self.split_dimension:
+                # Preserve the numbering of the unsplit entities
+                facets = chain(top[dim].values(), facets)
+            unique_facets = dict.fromkeys(facets)
+            new_topology[dim] = dict(enumerate(unique_facets))
+        new_topology[sd] = dict(enumerate(simplices))
+
+        parent = ref_el if dimension == sd else PowellSabinSplit(ref_el, dimension=dimension+1)
+        super().__init__(parent, tuple(new_verts), new_topology)
 
     def construct_subcomplex(self, dimension):
-        """Constructs the reference subcomplex of the parent cell subentity
+        """Constructs the reference subcomplex of the parent complex
         specified by subcomplex dimension.
         """
         if dimension == self.get_dimension():
             return self
-        ref_el = self.construct_subelement(dimension)
-        if dimension == 0:
-            return ref_el
+        parent = self.get_parent_complex()
+        subcomplex = parent.construct_subcomplex(dimension)
+        if dimension < self.split_dimension:
+            return subcomplex
         else:
             # Powell-Sabin on facets is Powell-Sabin
-            return PowellSabinSplit(ref_el)
+            return PowellSabinSplit(subcomplex, dimension=self.split_dimension)
+
+
+class AlfeldSplit(PowellSabinSplit):
+    """Splits a simplicial complex by connecting cell vertices to their
+    barycenter.
+    """
+    def __init__(self, ref_el):
+        sd = ref_el.get_spatial_dimension()
+        super().__init__(ref_el, dimension=sd)
+
+
+class WorseyFarinSplit(PowellSabinSplit):
+    """Splits a simplicial complex by connecting cell and face vertices to their
+    barycenter.
+    """
+    def __init__(self, ref_el):
+        super().__init__(ref_el, dimension=2)
 
 
 class PowellSabin12Split(SplitSimplicialComplex):
```

Comment:
any chance of fusing this with Powell Sabin and other splits above?  I did this as a 1-off. but maybe it's not worth the effort to optimize since it works.

Reply:
This one is is a different beast